build/source/engine/read_force.f90

```diff
@@ -417,49 +214,313 @@ subroutine read_force(istep,iFile,iRead,ncid,time_data,forcStruct,err,message)
+ logical(lgt),parameter            :: checkTime=.false.  ! flag to check the time
+ real(dp)                          :: dsec             ! double precision seconds (not used)
+ real(dp)                          :: dataJulDay       ! julian day of current forcing data step being read
+ real(dp),dimension(nHRUlocal)     :: dataVec          ! vector of data
+ real(dp),dimension(1)             :: dataVal          ! single data value
+ real(dp),parameter                :: dataMin=-1._dp   ! minimum allowable data value (all forcing variables should be positive)
+ logical(lgt),dimension(size(forc_meta)) :: checkForce ! flags to check forcing data variables exist
+ logical(lgt),parameter            :: simultaneousRead=.true. ! flag to denote reading all HRUs at once
+ ! Start procedure here
+ err=0; message="readForcingData/"
+
+ ! initialize time and forcing data structures
+ time_data(:) = integerMissing
+
+ ! read time data from iRead location in netcdf file
+ err = nf90_inq_varid(ncid,'time',varId);                   if(err/=nf90_noerr)then; message=trim(message)//'trouble finding time variable/'//trim(nf90_strerror(err)); return; endif
+ err = nf90_get_var(ncid,varId,varTime,start=(/iRead/));    if(err/=nf90_noerr)then; message=trim(message)//'trouble reading time variable/'//trim(nf90_strerror(err)); return; endif
+
+ ! check that the computed julian day matches the time information in the NetCDF file
+ dataJulDay = varTime(1)/forcFileInfo(iFile)%convTime2Days + refJulday_data
+ if(abs(currentJulday - dataJulDay) > verySmall)then
+  write(message,'(a,f18.8,a,f18.8)') trim(message)//'date for time step: ',dataJulDay,' differs from the expected date: ',currentJulDay
+  err=40; return
+ end if
+
+ ! convert julian day to time vector
+ ! NOTE: use small offset to force ih=0 at the start of the day
+ call compcalday(dataJulDay+smallOffset,         & ! input  = julian day
+                 time_data(iLookTIME%iyyy),      & ! output = year
+                 time_data(iLookTIME%im),        & ! output = month
+                 time_data(iLookTIME%id),        & ! output = day
+                 time_data(iLookTIME%ih),        & ! output = hour
+                 time_data(iLookTIME%imin),dsec, & ! output = minute/second
+                 err,cmessage)                     ! output = error control
+ if(err/=0)then; message=trim(message)//trim(cmessage); return; end if
+
+ ! check to see if any of the time data is missing
+ if(any(time_data(:)==integerMissing))then
+  do iline=1,size(time_data)
+   if(time_data(iline)==integerMissing)then; err=40; message=trim(message)//"variableMissing[var='"//trim(time_meta(iline)%varname)//"']"; return; end if
+  end do
+ end if
+
+ ! initialize flags for forcing data
+ checkForce(:) = .false.
+ checkForce(iLookFORCE%time) = .true.  ! time is handled separately
+
+ ! loop through forcing data variables
+ do iNC=1,forcFileInfo(iFile)%nVars
```

Comment:
Same as above - would it be faster if you only look for the data that you know you need? Rather than checking all the variables in the file?

Reply:
We only read the data that we need.

Reply:
The very next line of code skips variables that are not needed:
```fortran
! loop through forcing data variables
 do iNC=1,forcFileInfo(iFile)%nVars

  ! check variable is desired
  if(forcFileInfo(iFile)%var_ix(iNC)==integerMissing) cycle
```

Reply:
It could be even more clear if we combined these statements into one line in a `do concurrent` but I need to check that there are no dependencies among different loop iterations. Leaving as is for now.

```fortran
do concurrent(iNC=1:forcFileInfo(iFile)%nVars, forcFileInfo(iFile)%var_ix(iNC)/=integerMissing)
```